crates/sui-core/src/authority.rs

```diff
@@ -3158,8 +3158,22 @@ impl AuthorityState {
             "Creating advance epoch transaction"
         );
 
+        fail_point_async!("change_epoch_tx_delay");
         let _tx_lock = epoch_store.acquire_tx_lock(tx_digest).await;
 
+        // The tx could have been executed by state sync already - if so simply return an error.
+        // The checkpoint builder will shortly be terminated by reconfiguration anyway.
+        if self
+            .database
+            .is_tx_already_executed(tx_digest)
```

Comment:
Shouldn't we be checking this after holding the lock?

Reply:
this is after the lock, `_tx_lock` specifically. i don't think we need to wait until we have the execution_guard